### PR TITLE
Fix Gradient for trajectory

### PR DESCRIPTION
**Pull request by chaithyagr**
I have updated the tests for the gradient and updated the issue with the gradient for the trajectory

### CODE REVIEW

**Comment by chaithyagr** [comment]
I have made the tests a bit more stricter also, looks like we were very loose with the tests

File: tfkbnufft/tests/ndft_test.py [hunk 1]
```diff
@@ -25,29 +25,44 @@ def test_adjoint_and_gradients(im_size):
...
 
     # Test gradients with respect to trajectory location
     gradient_ndft_traj = g.gradient(I_ndft, ktraj)[0]
     gradient_nufft_traj = g.gradient(I_nufft, ktraj)[0]
-    tf_test.assertAllClose(gradient_ndft_traj, gradient_nufft_traj, atol=1e-2)
+    tf_test.assertAllClose(gradient_ndft_traj, gradient_nufft_traj, atol=1e-3)
+
+    dy_ndft = g.gradient(loss_ndft, I_ndft)
+    dy_nufft = g.gradient(loss_nufft, I_nufft)
+    tf_test.assertAllClose(dy_ndft, dy_nufft, atol=1e-4)
```
Comment (zaccharieramzi) [inline tfkbnufft/tests/ndft_test.py:59]: Why add these? I feel like those only test the gradient of the mse

File: tfkbnufft/tests/ndft_test.py [hunk 2]
```diff
@@ -25,29 +25,44 @@ def test_adjoint_and_gradients(im_size):
...
+
+    # Test gradients with respect to kdata
+    gradient_ndft_loss = g.gradient(loss_ndft, ktraj)[0]
+    gradient_nufft_loss = g.gradient(loss_nufft, ktraj)[0]
+    tf_test.assertAllClose(gradient_ndft_loss, gradient_nufft_loss, atol=5e-4)
```
Comment (zaccharieramzi) [inline tfkbnufft/tests/ndft_test.py:64]: same for those.

Generally speaking, I think the loss is not necessary here, why did you add it?

File: tfkbnufft/tests/ndft_test.py [hunk 3]
```diff
@@ -60,27 +75,45 @@ def test_forward_and_gradients(im_size):
...
     # This is gradient of NDFT from matrix, will help in debug
-    # gradient_ndft_matrix = -2j * np.pi * tf.transpose(tf.matmul(A, tf.transpose(tf.cast(r, tf.complex64) * tf.reshape(signal[0][0], (N*N,)))))
+    # gradient_ndft_matrix = -1j * tf.transpose(tf.matmul(A, tf.transpose(tf.cast(grid_r, tf.complex64) * signal[0][0])))
+
+tf.config.run_functions_eagerly(True)
+test_forward_and_gradients((10, 10))
+test_adjoint_and_gradients((10, 10))
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply tfkbnufft/tests/ndft_test.py:119]: also why add these function calls at the end?
These should be run by `pytest`

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
Very cool that you were able to identify this bug.
I would like to know more about it.

Also some questions re tests

File: tfkbnufft/tests/ndft_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply tfkbnufft/tests/ndft_test.py:119]: Ahh no, by mistake, was debugging locally

File: tfkbnufft/kbnufft.py [hunk 4]
```diff
@@ -189,7 +189,7 @@ def grad(dy):
                 grid_r = tf.cast(tf.meshgrid(*r, indexing='ij'), x.dtype)[None, ...]
                 fft_dx_dom = scale_and_fft_on_image_volume(
                 x * grid_r, scaling_coef, grid_size, im_size, norm, im_rank=im_rank)
-                dy_dom = tf.cast(-1j * dy * kbinterp(fft_dx_dom, om, interpob), tf.float32)
+                dy_dom = tf.cast(-1j * tf.math.conj(dy) * kbinterp(fft_dx_dom, om, interpob), tf.float32)
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply tfkbnufft/kbnufft.py:192]: Haha, its verrrry blurry even to me, maybe we discuss it off hand?

File: tfkbnufft/tests/ndft_test.py (hunk 1 above)
Comment (chaithyagr) [reply tfkbnufft/tests/ndft_test.py:59]: Yes you are right, again was in debug, will remove...

File: tfkbnufft/tests/ndft_test.py (hunk 2 above)
Comment (chaithyagr) [reply tfkbnufft/tests/ndft_test.py:60]: Ahhh yes, now this is needed! 
The Jacobians were right earlier, the issue was when it was applied to a dy with chain rule!

Which is why I had to add this as a part of the test.
If the above changes in module is not done, the NDFT and NUFFT gradinets are way off here, while the usual earlier gradients in test all are equal and the tests will convientently pass..

Comment (zaccharieramzi) [reply tfkbnufft/tests/ndft_test.py:60]: ah ok makes total sense

File: tfkbnufft/tests/ndft_test.py [hunk 5]
```diff
@@ -60,27 +71,37 @@ def test_forward_and_gradients(im_size):
...
-    tf_test.assertAllClose(gradient_ndft_kdata, gradient_nufft_kdata, atol=1e-2)
+    tf_test.assertAllClose(gradient_ndft_kdata, gradient_nufft_kdata, atol=5e-3)
 
     # Test gradients with respect to trajectory location
     gradient_ndft_traj = g.gradient(kdata_ndft, ktraj)[0]
     gradient_nufft_traj = g.gradient(kdata_nufft, ktraj)[0]
-    tf_test.assertAllClose(gradient_ndft_traj, gradient_nufft_traj, atol=1e-2)
+    tf_test.assertAllClose(gradient_ndft_traj, gradient_nufft_traj, atol=5e-3)
+
+    # Test gradients with respect to kdata
```
Comment (zaccharieramzi) [inline tfkbnufft/tests/ndft_test.py:102]:
```suggestion
    # Test gradients in chain rule with respect to ktraj
```

File: tfkbnufft/tests/ndft_test.py [hunk 6]
```diff
@@ -25,29 +25,40 @@ def test_adjoint_and_gradients(im_size):
...
-    tf_test.assertAllClose(gradient_ndft_kdata, gradient_nufft_kdata, atol=1e-2)
+    tf_test.assertAllClose(gradient_ndft_kdata, gradient_nufft_kdata, atol=5e-3)
 
     # Test gradients with respect to trajectory location
     gradient_ndft_traj = g.gradient(I_ndft, ktraj)[0]
     gradient_nufft_traj = g.gradient(I_nufft, ktraj)[0]
-    tf_test.assertAllClose(gradient_ndft_traj, gradient_nufft_traj, atol=1e-2)
+    tf_test.assertAllClose(gradient_ndft_traj, gradient_nufft_traj, atol=5e-3)
+
+    # Test gradients with respect to kdata
```
Comment (zaccharieramzi) [inline tfkbnufft/tests/ndft_test.py:57]:
```suggestion
    # Test gradients in chain rule with respect to ktraj
```

File: tfkbnufft/kbnufft.py [hunk 7]
```diff
@@ -43,6 +43,9 @@ def __init__(self, im_size, grid_size=None, numpoints=6, n_shift=None,
         self.im_size = im_size
         self.im_rank = len(im_size)
         self.grad_traj = grad_traj
+        if self.grad_traj:
+            warnings.warn('The gradient w.r.t trajectory is Experimental and WIP. '
```
Comment (zaccharieramzi) [inline tfkbnufft/kbnufft.py:47]: can you also add something in the readme?

File: README.md [hunk 8]
```diff
@@ -32,6 +32,16 @@ python profile_tfkbnufft.py
 
 These numbers were obtained with a Quadro P5000.
 
+
+## Gradients
+
+### w.r.t trajectory
+
+This is experimental currently and is WIP. Please be cautiously.
```
Comment (zaccharieramzi) [inline README.md:40]:
```suggestion
This is experimental currently and is WIP. Please be cautious. 
```

**Review by zaccharieramzi — APPROVED** [review]

**Comment by zaccharieramzi** [comment]
LGTM
